out_transitions for location with no transitions

Here is a small example of an automaton where the second location has no outgoing transition.
```julia
using HybridSystems

automaton = LightAutomaton(2)
add_transition!(automaton, 1, 1, 1)
add_transition!(automaton, 1, 2, 2)
m1 = "mode1"
m2 = "mode2"
modes = [m1, m2]
resetmaps = ["transition1", "transition2"]
switchings = [AutonomousSwitching() for _ in 1:2]
H = HybridSystem(automaton, modes, resetmaps, switchings)

println("outgoing transitions of location 1:")
out_transitions(H, 1)
println("outgoing transitions of location 2:")
out_transitions(H, 2)
```
This script breaks in the last line with the following message:
```
ERROR: LoadError: AssertionError: 1 <= r <= nstates(A)
```
After adding `println("q = $q, r = $r")` to [this line](https://github.com/blegat/HybridSystems.jl/blob/06b6f8050555e5f001c93bf3f0200a7fa5db12a7/src/light.jl#L111) I get this output:
```
q = 1, r = 1
q = 1, r = 2
outgoing transitions of location 1:
q = 1, r = 1
outgoing transitions of location 2:
q = 2, r = 0
ERROR: LoadError: AssertionError: 1 <= r <= nstates(A)
```
It seems that the graph method [here](https://github.com/blegat/HybridSystems.jl/blob/06b6f8050555e5f001c93bf3f0200a7fa5db12a7/src/light.jl#L175) returns `0` if there is no outgoing edge, which is apparently not expected by the other function.

--- a/src/light.jl
+++ b/src/light.jl
@@ -74,8 +74,9 @@ struct LightTransitionIterator{GT, ET, VT}
 end
 Base.eltype(::LightTransitionIterator{GT, ET}) where {GT, ET} = LightTransition{ET}
 function Base.length(tit::LightTransitionIterator)
-    return sum(edge -> length(tit.automaton.Σ[edge]),
-               tit.edge_iterator)
+    eit = tit.edge_iterator
+    # empty iterator must be handled separately (see #29)
+    return isempty(eit) ? 0 : sum(edge -> length(tit.automaton.Σ[edge]), eit)
 end
 function new_id_iterate(tit::LightTransitionIterator, edge, edge_state, ids, ::Nothing)
     return new_edge_iterate(tit, iterate(tit.edge_iterator, edge_state))
@@ -107,10 +108,14 @@ function transitions(A::LightAutomaton)
 end
 ntransitions(A::LightAutomaton) = A.nt
 
+function edge_object_no_assertion(A::LightAutomaton, q, r)
+    return LightGraphs.Edge(q, r)
+end
+
 function edge_object(A::LightAutomaton, q, r)
     @assert 1 <= q <= nstates(A)
     @assert 1 <= r <= nstates(A)
-    return LightGraphs.Edge(q, r)
+    return edge_object_no_assertion(A, q, r)
 end
 
 function add_transition!(A::LightAutomaton, q, r, σ)
@@ -166,13 +171,16 @@ event(A::LightAutomaton, t::LightTransition) = A.Σ[t.edge][t.id]
 target(::LightAutomaton, t::LightTransition) = t.edge.dst
 
 function in_transitions(A::LightAutomaton, s)
-    edges = MappedArrays.mappedarray(src -> edge_object(A, src, s),
-                                     LightGraphs.inneighbors(A.G, s))
+    it = LightGraphs.inneighbors(A.G, s)
+    # MappedArrays.mappedarray over empty array returns 'zero' (see #29)
+    edge_function = isempty(it) ? edge_object_no_assertion : edge_object
+    edges = MappedArrays.mappedarray(src -> edge_function(A, src, s), it)
     LightTransitionIterator(A, edges)
 end
 function out_transitions(A::LightAutomaton, s)
-    edges = MappedArrays.mappedarray(dst -> edge_object(A, s, dst),
-                                     LightGraphs.outneighbors(A.G, s))
+    it = LightGraphs.outneighbors(A.G, s)
+    edge_function = isempty(it) ? edge_object_no_assertion : edge_object
+    edges = MappedArrays.mappedarray(dst -> edge_function(A, s, dst), it)
     LightTransitionIterator(A, edges)
 end
 

--- a/src/light.jl
+++ b/src/light.jl
@@ -1,6 +1,8 @@
 import MappedArrays
 import LightGraphs
 
+using MappedArrays: ReadonlyMappedArray
+
 """
     LightAutomaton{GT, ET} <: AbstractAutomaton
 
@@ -167,15 +169,15 @@ event(A::LightAutomaton, t::LightTransition) = A.Σ[t.edge][t.id]
 target(::LightAutomaton, t::LightTransition) = t.edge.dst
 
 function in_transitions(A::LightAutomaton{GT, ET}, s) where {GT, ET}
-    edges = MappedArrays.mappedarray(ET,
-                                     src -> edge_object(A, src, s),
-                                     LightGraphs.inneighbors(A.G, s))
+    f(src) = edge_object(A, src, s)
+    inneigh = LightGraphs.inneighbors(A.G, s)
+    edges = ReadonlyMappedArray{ET, 1, typeof(inneigh), typeof(f)}(f, inneigh)
     LightTransitionIterator(A, edges)
 end
 function out_transitions(A::LightAutomaton{GT, ET}, s) where {GT, ET}
-    edges = MappedArrays.mappedarray(ET,
-                                     dst -> edge_object(A, s, dst),
-                                     LightGraphs.outneighbors(A.G, s))
+    f(dst) = edge_object(A, s, dst)
+    outneigh = LightGraphs.outneighbors(A.G, s)
+    edges = ReadonlyMappedArray{ET, 1, typeof(outneigh), typeof(f)}(f, outneigh)
     LightTransitionIterator(A, edges)
 end
 

--- a/src/light.jl
+++ b/src/light.jl
@@ -108,14 +108,10 @@ function transitions(A::LightAutomaton)
 end
 ntransitions(A::LightAutomaton) = A.nt
 
-function edge_object_no_assertion(A::LightAutomaton, q, r)
-    return LightGraphs.Edge(q, r)
-end
-
 function edge_object(A::LightAutomaton, q, r)
     @assert 1 <= q <= nstates(A)
     @assert 1 <= r <= nstates(A)
-    return edge_object_no_assertion(A, q, r)
+    return LightGraphs.Edge(q, r)
 end
 
 function add_transition!(A::LightAutomaton, q, r, σ)
@@ -170,17 +166,16 @@ source(::LightAutomaton, t::LightTransition) = t.edge.src
 event(A::LightAutomaton, t::LightTransition) = A.Σ[t.edge][t.id]
 target(::LightAutomaton, t::LightTransition) = t.edge.dst
 
-function in_transitions(A::LightAutomaton, s)
-    it = LightGraphs.inneighbors(A.G, s)
-    # MappedArrays.mappedarray over empty array returns 'zero' (see #29)
-    edge_function = isempty(it) ? edge_object_no_assertion : edge_object
-    edges = MappedArrays.mappedarray(src -> edge_function(A, src, s), it)
+function in_transitions(A::LightAutomaton{GT, ET}, s) where {GT, ET}
+    edges = MappedArrays.mappedarray(ET,
+                                     src -> edge_object(A, src, s),
+                                     LightGraphs.inneighbors(A.G, s))
     LightTransitionIterator(A, edges)
 end
-function out_transitions(A::LightAutomaton, s)
-    it = LightGraphs.outneighbors(A.G, s)
-    edge_function = isempty(it) ? edge_object_no_assertion : edge_object
-    edges = MappedArrays.mappedarray(dst -> edge_function(A, s, dst), it)
+function out_transitions(A::LightAutomaton{GT, ET}, s) where {GT, ET}
+    edges = MappedArrays.mappedarray(ET,
+                                     dst -> edge_object(A, s, dst),
+                                     LightGraphs.outneighbors(A.G, s))
     LightTransitionIterator(A, edges)
 end
 

--- a/test/automata.jl
+++ b/test/automata.jl
@@ -68,9 +68,17 @@ end
         t1 = add_transition!(automaton, 1, 2, 1)
         @test has_transition(automaton, 1, 2)
         @test !has_transition(automaton, 1, 1)
+        @test length(out_transitions(automaton, 1)) == 1
+        @test length(out_transitions(automaton, 2)) == 0
+        @test length(in_transitions(automaton, 1)) == 0
+        @test length(in_transitions(automaton, 2)) == 1
         t2 = add_transition!(automaton, 1, 1, 1)
         t3 = add_transition!(automaton, 2, 1, 2)
         @test !has_transition(automaton, 2, 2)
+        @test length(out_transitions(automaton, 1)) == 2
+        @test length(out_transitions(automaton, 2)) == 1
+        @test length(in_transitions(automaton, 1)) == 2
+        @test length(in_transitions(automaton, 2)) == 1
         t4 = add_transition!(automaton, 2, 2, 1)
         t5 = add_transition!(automaton, 1, 2, 2)
         @test has_transition(automaton, t1)

